engineering/backend/restful.md

```diff
@@ -33,27 +33,65 @@ Always version your API. Version via the URL, not via headers. Versioning APIs a
 ## Filter, Search and Sort
 Use query parameters for advanced filtering, sorting & searching.
 
-Pagination:
+###Pagination
+
+**Offset pagination:**
+
+- `page` - current page wanted to fetch
+- `size` - number of items per page
+
+- The API should return `totalItems` and `totalPages`
+
+Example:
 ```
 GET /companies?page=23&size=50
 ```
 
-Sort:
+**Cursor pagination**
+
+- `cursor` - is the reference element in the list. Best brattice for choosing the cursor is only sort by unique sequential column, create new column which is concatenate columns if you have to order by multiple none unique columns.
+- `size` - number of items per page
+
+- The API should return `previousCursor` and `nextCursor`
+
 ```
-GET /companies?sort=-created_at
+GET /companies?cursor=12&size=50&cursorDirection=
 ```
 
-Filter:
+### Sort:
+
+- Validate all sort able fields
+- By default sort in ascending order
+- Use  `minus(-)` prefix to indicate sort in descending order
+
+Example
 ```
-GET /companies?category=software&location=saigon|hanoi
+# sort by created_at in descending order
+GET /companies?sort=-created_at
+
+# sort by created_at in ascending order
+GET /companies?sort=created_at
 ```
 
-Filter by time range:
+### Filter:
+
+- Validate all enums values
+- Use repeated query keys for multiple values
+- Use RFC339 time format or unix timestamp for date time filter 
+
 ```
-GET /companies?time=fromTime-toTime
+# filter by category
+GET /companies?category=software
+
+# filter by locations
+GET /companies?location=saigon&location=hanoi
+
+# Filter by time range
```

Comment:
It makes me confused about the markdown convention. As we knew, One # character for Heading 1. We should update the way we index the headings.
```suggestion
#### Filter by category
GET /companies?category=software

#### Filter by locations
GET /companies?location=saigon&location=hanoi

#### Filter by time range
```